### PR TITLE
Support Debug/Release build configurations

### LINKED ISSUE #185: Need DEBUG or RELEASE command line option for make
Adding an option to the make commands to enable optimizations and disable debug symbols to help code size.  Something like 

make BSP=metal PROGRAM=hello TARGET=sifive-hifive1 *RELEASE=on* software 

If RELEASE is not specified have it default to no optimizations to allow for better debugging capabilities.  This can be tied in with the Eclipse profiles for 'Release' and 'Debug' target options

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Closes #185

### CODE REVIEW

**Comment by bsousi5** [comment]
Since majority of the flags are the same, would it make more sense to only have one .mk, have it include the other one plus the different. Otherwise two files need to be maintained.

File: Makefile [hunk 1]
```diff
@@ -25,8 +25,8 @@ PROGRAM_ROOT ?= $(abspath .)
 
 SRC_DIR = $(PROGRAM_ROOT)/software/$(PROGRAM)
 
-PROGRAM_ELF = $(SRC_DIR)/$(PROGRAM)
-PROGRAM_HEX = $(SRC_DIR)/$(PROGRAM).hex
+PROGRAM_ELF = $(SRC_DIR)/$(CONFIGURATION)/$(PROGRAM)
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply Makefile]: I can add that.

File: Makefile [hunk 2]
```diff
@@ -64,31 +64,34 @@ help:
...
+	@echo "    Builds the requested PROGRAM for the TARGET using the"
+	@echo "    specified build CONFIGURATION."
 	@echo ""
 	@echo " metal [TARGET=$(TARGET)]"
-	@echo "    Build the Freedom Metal library for TARGET"
+	@echo "    Builds the Freedom Metal library for TARGET."
 	@echo ""
-	@echo " clean [PROGRAM=$(PROGRAM) TARGET=$(TARGET)]:"
-	@echo "    Clean compiled objects for a specified "
+	@echo " clean [PROGRAM=$(PROGRAM)] [TARGET=$(TARGET)]:"
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply Makefile]: I'll add that to the help text.

File: Makefile [hunk 3]
```diff
@@ -64,31 +64,34 @@ help:
 	@echo " SiFive Freedom E Software Development Kit "
 	@echo " Makefile targets:"
 	@echo ""
-	@echo " software [PROGRAM=$(PROGRAM) TARGET=$(TARGET)]:"
-	@echo "    Build a software program to load with the"
-	@echo "    debugger."
+	@echo " software [PROGRAM=$(PROGRAM)] [TARGET=$(TARGET)]"
+	@echo "          [CONFIGURATION=$(CONFIGURATION)]:"
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply Makefile]: The default is Debug.

**Comment by nategraff-sifive** [comment]
@cgsfv 

- The Metal library build is also modified by CONFIGURATION.
- When the Metal sources are included in the standalone project, it is built by the all target.
- New configurations are as simple as creating new Name.mk files in the top level directory and then just requesting CONFIGURATION=Name
- I don't like creating a custom file extension just for Makefiles. Especially for people using the command line, the text editor won't necessary choose the right syntax highlighting.

**Comment by nategraff-sifive** [comment]
I'm also tempted to get rid of the camel-case configuration names. Everything else is lower-case, so I think that should be better.

File: doc/sphinx/userguide/standalone.rst [hunk 4]
```diff
@@ -15,5 +15,5 @@ be included in the generated project as a pre-built archive.
 
 .. code-block:: bash
 
-   make [PROGRAM=hello] [TARGET=sifive-hifive1] [INCLUDE_METAL_SOURCES=1] STANDALONE_DEST=/path/to/desired/location standalone
+   make [PROGRAM=hello] [TARGET=sifive-hifive1] [CONFIGURATION=Debug] [INCLUDE_METAL_SOURCES=1] STANDALONE_DEST=/path/to/desired/location standalone
```
Comment (cgsfv) [inline doc/sphinx/userguide/standalone.rst]: You missed this Camel case :-)

Comment (nategraff-sifive) [reply doc/sphinx/userguide/standalone.rst]: Thanks for the catch :)

**Comment by nategraff-sifive** [comment]
Hmm, yeah, the Metal library doesn't get built into a separate directory. Let me see if it's possible to change that without disrupting too much.

**Comment by nategraff-sifive** [comment]
@cgsfv Metal is now being built into bsp/TARGET/install/CONFIGURATION.

I'm getting rid of INCLUDE_METAL_SOURCES, because now that the build flags change it doesn't make sense to have a prebuilt Metal library. Standalone projects will just rebuild the library.

**Comment by sifivekevin** [comment]
OK.  So no more INCLUDE_METAL_SOURCES. I'll need to remove that option from
the project wizard.

File: scripts/standalone.mk [hunk 5]
```diff
@@ -72,17 +77,49 @@ endif
...
+$(error Unable to find the Makefile $(CONFIGURATION).mk for CONFIGURATION=$(CONFIGURATION))
+endif
+include $(CONFIGURATION).mk
+
 #############################################################
 # Software
 #############################################################
 
-PROGRAM_ELF ?= $(SRC_DIR)/$(PROGRAM)
+PROGRAM_ELF ?= $(SRC_DIR)/$(CONFIGURATION)/$(PROGRAM)
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply scripts/standalone.mk]: Command line users have been expecting the elf to not have the extension. I felt like this was a good compromise.

File: scripts/standalone.mk [hunk 6]
```diff
@@ -72,17 +77,49 @@ endif
...
+#############################################################
+
+# Set the arch, ABI, and code model
+RISCV_CFLAGS   += -march=$(RISCV_ARCH) -mabi=$(RISCV_ABI) -mcmodel=$(RISCV_CMODEL)
+RISCV_CXXFLAGS += -march=$(RISCV_ARCH) -mabi=$(RISCV_ABI) -mcmodel=$(RISCV_CMODEL)
+# Prune unused functions and data
+RISCV_CFLAGS   += -ffunction-sections -fdata-sections
+RISCV_CXXFLAGS += -ffunction-sections -fdata-sections
+# Include the Metal headers
+RISCV_CFLAGS   += -I$(abspath $(BSP_DIR)/install/$(CONFIGURATION)/include/)
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply scripts/standalone.mk]: Headers are now excluded, only the built libs get special treatment using the --libdir configure flag.

File: scripts/standalone.mk [hunk 7]
```diff
@@ -95,21 +132,26 @@ PROGRAM_SRCS = $(wildcard $(SRC_DIR)/*.c) $(wildcard $(SRC_DIR)/*.h) $(wildcard
 
 $(PROGRAM_ELF): \
 		$(PROGRAM_SRCS) \
-		$(BSP_DIR)/install/lib/libmetal.a \
-		$(BSP_DIR)/install/lib/libmetal-gloss.a \
+		$(BSP_DIR)/install/$(CONFIGURATION)/lib/libmetal.a \
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply scripts/standalone.mk]: Swapped.

File: scripts/standalone.mk (hunk 5 above)
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply scripts/standalone.mk]: Maybe this is a bad argument. We already changed the path...

[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply scripts/standalone.mk]: I see it your way, done.

File: scripts/libmetal.mk [hunk 8]
```diff
@@ -25,21 +30,21 @@ $(BSP_DIR)/build/Makefile:
...
-$(BSP_DIR)/install/lib/libmetal.a: $(BSP_DIR)/install/lib/libriscv__mmachine__$(TARGET).a
+$(METAL_LIB_DIR)/libmetal.a: $(METAL_LIB_DIR)/libriscv__mmachine__$(TARGET).a
 	cp $< $@
 
-$(BSP_DIR)/install/lib/libmetal-gloss.a: $(BSP_DIR)/install/lib/libriscv__menv__metal.a
+$(METAL_LIB_DIR)/libmetal-gloss.a: $(METAL_LIB_DIR)/libriscv__menv__metal.a
 	cp $< $@
 
 .PHONY: clean-metal
 clean-metal:
```
Comment (cgsfv) [inline scripts/libmetal.mk]: The metal-clean target removes all CONFIGURATIONS - is this on purpose?
I would like it only to remove a single CONFIGURATION, just like the clean-software target does.

Comment (nategraff-sifive) [reply scripts/libmetal.mk]: So yeah, right now it's removing the install directory entirely.

For removing a single configuration, it probably makes most sense to remove install/lib/CONFIGURATION. The question then is: when do we remove the installed Metal headers? Do we just leave them, or do we add a check and if we're removing the last config, then also delete the headers?

Comment (nategraff-sifive) [reply scripts/libmetal.mk]: I pushed a commit to do the latter (remove the headers if we're removing the last CONFIGURATION).

**Pull request merged**